typing/gprinttyp.ml

```diff
@@ -0,0 +1,872 @@
+        in
+        { dg with elts = Elt_map.add entry d dg.elts }
+    | None ->
+        let elts = Elt_map.add entry d dg.elts in
+        { elts; graph = add_to_subgraph dg.graph entry }
+
+  let rec hyperedges_of_memo ty params id abbrev dg =
+    match abbrev with
+    | Types.Mnil -> dg
+    | Types.Mcons (_priv, _p, t1, t2, rem) ->
+        let s, dg = ty params t1 dg in
+        let exp, dg = ty params t2 dg in
+        dg |>
+        add memo
+          (Hyperedge
+             [From, dotted, id;
+              Toward, dotted, s;
+              Toward, Decoration.make [txt "expand"], exp
+             ])
+        |> hyperedges_of_memo ty params id rem
+    | Types.Mlink rem -> hyperedges_of_memo ty params id !rem dg
+
+  let rec edges_of_memo ty params abbrev dg =
+    match abbrev with
+    | Types.Mnil -> dg
+    | Types.Mcons (_priv, _p, t1, t2, rem) ->
+        let x, dg = ty params t1 dg in
+        let y, dg = ty params t2 dg in
+        dg |> add memo (Edge (x,y)) |> edges_of_memo ty params rem
+    | Types.Mlink rem -> edges_of_memo ty params !rem dg
+
+  let expansions ty params id memo dg =
+    if params.expansion_as_hyperedge then
+      hyperedges_of_memo ty params id memo dg
+    else
+      edges_of_memo ty params memo dg
+
+  let labelk k fmt = kasprintf (fun s -> k  [txt s]) fmt
+  let labelf fmt = labelk Fun.id fmt
+  let labelr fmt = labelk Decoration.make fmt
+
+  let add_node explicit_d color id tynode dg =
+    let d = labelf "<SUB>%a</SUB>" Pp.prettier_index id in
+    let d = match color with
+    | None -> Decoration.make d
+    | Some x -> Decoration.(make (filled x :: d))
+    in
+    let d = Decoration.merge explicit_d d in
+    add d tynode dg
+
+  let field_node color lbl rf =
+    let col = match color with
+      | None -> []
+      | Some c -> [Decoration.Color c]
+    in
+    let pr_lbl ppf = match lbl with
+      | None -> ()
+      | Some lbl -> fprintf ppf "`%s" lbl
+    in
+    let lbl =
+      Types.match_row_field
+        ~absent:(fun _ -> labelf "`-%t" pr_lbl)
+        ~present:(fun _ -> labelf "&gt;%t" pr_lbl)
+        ~either:(fun c _tl m _e ->
+            labelf "%s%t%s"
+              (if m then "?" else "")
+              pr_lbl
+              (if c then "(∅)" else "")
+          )
+        rf
+    in
+    Decoration.(make (Shape Diamond::col@lbl))
+
+  let group ty id0 lbl l dg =
+    match l with
+    | [] -> dg
+    | first :: l ->
+      let sub = { dg with graph = empty_subgraph } in
+      let id, sub = ty first sub in
+      let sub = List.fold_left (fun dg t -> snd (ty t dg)) sub l in
+      let dg = { sub with graph = add_subgraph (lbl,sub.graph) dg.graph } in
+      dg |> add std (Edge(id0,id))
+
+  let split_fresh_typ params ty0 g =
+    let (id, color, desc) = Index.split params ty0 in
+    let tynode = Node id in
+    if Elt_map.mem tynode g then id, None else id, Some (tynode,color,desc)
+
+  let rec inject_typ ?(follow_expansions=true) params ty0 dg =
```

Comment:
Why is  `follow_expansions` not part of `params`?

Reply:
Mostly because I was thinking of it as an internal parameter for the `group` function, but indeed there is no reason to not expose it as a parameter. I have also added an `update_params` function to be able to tweak fields of parameters more easily.